x-pack/plugins/observability_solution/observability/server/services/slo/transform_generators/synthetics_availability.ts

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { TransformPutTransformRequest } from '@elastic/elasticsearch/lib/api/types';
+import { estypes } from '@elastic/elasticsearch';
+import {
+  ALL_VALUE,
+  syntheticsAvailabilityIndicatorSchema,
+  timeslicesBudgetingMethodSchema,
+} from '@kbn/slo-schema';
+import { getElasticsearchQueryOrThrow, TransformGenerator } from '.';
+import {
+  getSLOTransformId,
+  SLO_DESTINATION_INDEX_NAME,
+  SLO_INGEST_PIPELINE_NAME,
+} from '../../../../common/slo/constants';
+import { getSLOTransformTemplate } from '../../../assets/transform_templates/slo_transform_template';
+import { SyntheticsAvailabilityIndicator, SLO } from '../../../domain/models';
+import { InvalidTransformError } from '../../../errors';
+export class SyntheticsAvailabilityTransformGenerator extends TransformGenerator {
+  public getTransformParams(slo: SLO, spaceId: string): TransformPutTransformRequest {
+    if (!syntheticsAvailabilityIndicatorSchema.is(slo.indicator)) {
+      throw new InvalidTransformError(`Cannot handle SLO of indicator type: ${slo.indicator.type}`);
+    }
+
+    return getSLOTransformTemplate(
+      this.buildTransformId(slo),
+      this.buildDescription(slo),
+      this.buildSource(slo, slo.indicator, spaceId),
+      this.buildDestination(),
+      this.buildGroupBy(slo, slo.indicator),
+      this.buildAggregations(slo),
+      this.buildSettings(slo, 'event.ingested')
+    );
+  }
+
+  private buildTransformId(slo: SLO): string {
+    return getSLOTransformId(slo.id, slo.revision);
+  }
+
+  private buildGroupBy(slo: SLO, indicator: SyntheticsAvailabilityIndicator) {
+    const hasTags =
+      !indicator.params.tags?.find((param) => param.value === ALL_VALUE) &&
+      indicator.params.tags?.length;
+    const hasProjects =
+      !indicator.params.projects?.find((param) => param.value === ALL_VALUE) &&
+      indicator.params.projects?.length;
+    // These groupBy fields must match the fields from the source query, otherwise
+    // the transform will create permutations for each value present in the source.
+    // E.g. if environment is not specified in the source query, but we include it in the groupBy,
+    // we'll output documents for each environment value
+    const extraGroupByFields = {
+      'monitor.id': { terms: { field: 'monitor.id' } },
+      'observer.name': { terms: { field: 'observer.name' } },
+      config_id: { terms: { field: 'config_id' } },
+      ...(hasTags && {
+        tags: { terms: { field: 'tags' } },
+      }),
+      ...(hasProjects && {
+        'monitor.project.id': { terms: { field: 'monitor.project.id' } },
+      }),
+    };
```

Comment:
I'd like to double check the logic here since I'm not sure if the `monitor.id` term aggs should not be conditional to the ALL_VALUE as we do for tags and projects. 

**OR**, are we always grouping by monitor.id and observer.name because we always groupBy using `[monitor.name, observer.geo.name]` which are directly linked to monitor.id and observer.name ?
If that's the case, I think we should enforce `['monitor.name', 'observer.geo.name'];` groupings and not only fallback to these values, i.e. don't use `slo.groupBy` at all.

Also, what about config_id? Can't we merge two config_id for a same monitor_id,observer_name?

Reply:
I think you're right. The issue is that I continue to think about it from the perspective of always grouping by monitor and location, rather than accounting for custom group_by. Since they may not always be grouped by the individual monitor, I'll adjust this.

`observer.name` and `config_id` are fields that are required to link back to the Synthetics app. I use them on the details page to hyperlink the monitor name. I suppose, if they aren't using the default group by, this could once again cause issues.